util/messaging.py

```diff
@@ -0,0 +1,234 @@
+        while not coord.should_stop():
+            try:
+                content = session.run(self._inbound_dequeue)
+            except (tf.errors.CancelledError, tf.errors.OutOfRangeError):
+                # queue got closed
+                return
+            # de-pickle header and call arguments
+            args = (session,) + pickle.loads(content)
+            # Create a thread to decouple blocking cross-cluster call/response chains.
+            Thread(target=self._receive, args=args).start()
+
+    def start_queue_threads(self, session, coord):
+        '''
+        Starts required threads for receiving and sending messages.
+        'session' - The current session
+        'coord' - A thread coordinator to automatically stop the threads
+        '''
+        threads = []
+        for routine in [self._send_loop, self._receive_loop]:
+            thread = Thread(target=routine, args=(session,coord))
+            thread.daemon = True
+            thread.start()
+            threads.append(thread)
+        log.debug('Started queue threads')
+        return threads
+
+    def close_queues(self, session):
+        '''
+        Closes all queues of this instance.
+        'session' - The current session
+        '''
+        session.run(self._inbound_close)
+        for close in self._outbound_closes:
+            session.run(close)
+        log.debug('Closed queue threads')
+
+    def call_async(self, job, task, function, callback, *args):
+        '''
+        Performs an asynchronous (non-blocking) RPC call to a remote node's function.
+        If 'job' and 'task' address the local node, the call is performed locally.
+        'job' - Job name of the local cluster node
+        'task' - Task index of the local cluster node
+        'function' - Name of the remote function (member of class derived from ClusterMessagingClient)
+        'callback' - Will be called with return value tuple of remote function as arguments,
+            if RPC call succeeded. Omitting it (None) results in dependency free state
+            (the local client can safely be stopped/ended).
+            Can also be an Event object (used internally by 'call').
+        'args' - Arguments to the remote function.
+        Returns a callback id (0 in case of no callback).
+        '''
+        # target client id (shared_name for the apropos queue)
+        id = self._id_pattern % (job, task)
+        if self.id == id:
+            # local call - no queueing needed
+            log.step('Local call of function: %s, arguments: %r' % (function, args))
+            results = self._call(function, args)
+            if callback:
+                callback(results)
+            return -1
+        log.step('Calling worker: %s, function: %s, arguments: %r' % (id, function, args))
+        # default value in case of no callback, if not waiting for responseno
+        # and no response should be produced on the receiving end.
+        response_id = 0
+        if callback:
```

Comment:
Same question, shouldn't that be `if callable(callback)` ?

Reply:
No, as this is dealt with at a different location in the code.